Sources/SwiftCompilerPluginMessageHandling/JSON/JSONEncoding.swift

```diff
@@ -0,0 +1,605 @@
+    case .object(let dict): return dict.count
+    default: preconditionFailure("Count does not apply to scalar")
+    }
+  }
+
+  func set(key: String, value: JSONReference) {
+    guard case .object(var dict) = backing else {
+      preconditionFailure()
+    }
+    backing = .null
+    dict[key] = value
+    backing = .object(dict)
+  }
+
+  func append(_ value: JSONReference) {
+    guard case .array(var arr) = backing else {
+      preconditionFailure()
+    }
+    backing = .null
+    arr.append(value)
+    backing = .array(arr)
+  }
+
+  static let null: JSONReference = .init(backing: .null)
+  static let trueKeyword: JSONReference = .init(backing: .trueKeyword)
+  static let falseKeyword: JSONReference = .init(backing: .falseKeyword)
+
+  @inline(__always)
+  static func newArray() -> JSONReference {
+    .init(backing: .array([]))
+  }
+
+  @inline(__always)
+  static func newObject() -> JSONReference {
+    .init(backing: .object([:]))
+  }
+
+  @inline(__always)
+  static func string(_ str: String) -> JSONReference {
+    .init(backing: .string(str))
+  }
+  @inline(__always)
+  static func number(_ integer: some BinaryInteger & LosslessStringConvertible) -> JSONReference {
+    .init(backing: .number(String(integer)))
+  }
+  @inline(__always)
+  static func number(_ floating: some BinaryFloatingPoint & LosslessStringConvertible) -> JSONReference {
+    .init(backing: .number(String(floating)))
+  }
+}
+
+/// Serialize JSONReference to [UInt8] data.
+private struct JSONWriter {
+  var data: [UInt8]
+  init() {
+    data = []
+  }
+
+  mutating func write(ascii: UInt8) {
+    data.append(ascii)
+  }
+  mutating func write(scalar: UnicodeScalar) {
+    data.append(UInt8(ascii: scalar))
+  }
+  mutating func write(string: StaticString) {
+    string.withUTF8Buffer { buffer in
+      data.append(contentsOf: buffer)
+    }
+  }
+  mutating func write(utf8: some Collection<UInt8>) {
+    data.append(contentsOf: utf8)
+  }
+
+  mutating func serialize(value: JSONReference) {
+    switch value.backing {
+    case .null:
+      write(string: "null")
+    case .trueKeyword:
+      write(string: "true")
+    case .falseKeyword:
+      write(string: "false")
+    case .string(let string):
+      serialize(string: string)
+    case .number(var string):
+      string.withUTF8 {
+        write(utf8: $0)
+      }
+    case .array(let array):
+      serialize(array: array)
+    case .object(let dictionary):
+      serialize(object: dictionary)
+    }
+  }
+
+  mutating func serialize(string: String) {
+    var string = string
+    write(scalar: "\"")
+    string.withUTF8 { utf8 in
+      var cursor = utf8.baseAddress!
+      let end = utf8.baseAddress! + utf8.count
+      var mark = cursor
+      func flush() {
+        if cursor > mark {
+          write(utf8: UnsafeBufferPointer(start: mark, count: cursor - mark))
+        }
+        cursor += 1
+        mark = cursor
+      }
+      while cursor != end {
```

Comment:
I think instead of a `while` loop here and making sure that we do increment `cursor` in every case, this could just be `for byte in utf8`. Then you are also statically guaranteed to always make forward progress.

Reply:
We need `cursor` value inside the loop. maybe `for cursor in utf8.baseAddress!..<(utf8.baseAddress! + utf8.count)`

Reply:
Oh, I missed the usage of `cursor` directly. The new `for` loop is great 👍🏽